# 머신러닝 분류실습
- part2 ch2


In [20]:
import pandas as pd

train = pd.read_csv("../data/train.csv")
test = pd.read_csv("../data/test.csv")

In [3]:
train.head()

,id,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,3331,34.0,State-gov,177331,Some-college,10,Married-civ-spouse,Prof-specialty,Husband,Black,Male,4386,0,40.0,United-States,>50K
1,19749,58.0,Private,290661,HS-grad,9,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,40.0,United-States,<=50K
2,1157,48.0,Private,125933,Some-college,10,Widowed,Exec-managerial,Unmarried,Black,Female,0,1669,38.0,United-States,<=50K
3,693,58.0,Private,100313,Some-college,10,Married-civ-spouse,Protective-serv,Husband,White,Male,0,1902,40.0,United-States,>50K
4,12522,41.0,Private,195661,Some-college,10,Married-civ-spouse,Transport-moving,Husband,White,Male,0,0,54.0,United-States,<=50K


## EDA

In [5]:
train.describe(include = 'object')

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,27642,29304,29304,27636,29304,29304,29304,28767,29304
unique,8,16,7,14,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,20451,9449,13466,3683,11845,25022,19578,26240,22263


In [6]:
train.describe()

,id,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,29304.000000,29292.000000,2.930400e+04,29304.000000,29304.000000,29304.000000,29291.000000
mean,16264.027880,38.553223,1.897488e+05,10.080842,1093.858722,86.744506,40.434229
std,9384.518323,13.628811,1.055250e+05,2.570824,7477.435640,401.518928,12.324036
min,0.000000,-38.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,8145.750000,28.000000,1.177890e+05,9.000000,0.000000,0.000000,40.000000
50%,16253.500000,37.000000,1.783765e+05,10.000000,0.000000,0.000000,40.000000
75%,24374.250000,48.000000,2.370682e+05,12.000000,0.000000,0.000000,45.000000
max,32560.000000,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [7]:
train.isnull().sum()

id                   0
age                 12
workclass         1662
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1668
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week      13
native.country     537
income               0
dtype: int64

In [8]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29304 entries, 0 to 29303
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              29304 non-null  int64  
 1   age             29292 non-null  float64
 2   workclass       27642 non-null  object 
 3   fnlwgt          29304 non-null  int64  
 4   education       29304 non-null  object 
 5   education.num   29304 non-null  int64  
 6   marital.status  29304 non-null  object 
 7   occupation      27636 non-null  object 
 8   relationship    29304 non-null  object 
 9   race            29304 non-null  object 
 10  sex             29304 non-null  object 
 11  capital.gain    29304 non-null  int64  
 12  capital.loss    29304 non-null  int64  
 13  hours.per.week  29291 non-null  float64
 14  native.country  28767 non-null  object 
 15  income          29304 non-null  object 
dtypes: float64(2), int64(5), object(9)
memory usage: 3.6+ MB


In [9]:
## 각 target별 빈도수 -> value_counts()
train['income'].value_counts()

income
<=50K    22263
>50K      7041
Name: count, dtype: int64

In [12]:
train['age'].value_counts()

age
 36.0    801
 33.0    794
 34.0    793
 31.0    789
 30.0    789
        ... 
 85.0      2
-33.0      1
-38.0      1
 87.0      1
-25.0      1
Name: count, Length: 75, dtype: int64

## 데이터 전처리
- 클리닝, 인코딩, 스케일링

### 클리닝

In [ ]:
#1. 클리닝 (결측치, 이상치 처리)
  ## 결측치
df = train.dropna() # 하나라도 있으면 다 삭제
df2 = train.dropna(subset = ['workclass']) # 특정 컬럼에 결측치가 있을때 삭제
#print(df.isnull().sum())
print(df2.isnull().sum())

id                  0
age                12
workclass           0
fnlwgt              0
education           0
education.num       0
marital.status      0
occupation          6
relationship        0
race                0
sex                 0
capital.gain        0
capital.loss        0
hours.per.week     13
native.country    515
income              0
dtype: int64


#### Q. 결측치 채우는법
1. 처음엔 삭제해서 진행하고 
  2. 모델성능 확인함
3. 채우고
  4. 2차결과로 모델성능 확인함
 => 모델성능이 좋아지는 방향으로 바꿔해보기

In [ ]:
## 결측치  - 범주형은 fillna()

In [ ]:
## 이상치
  #(생략)

### 인코딩

In [21]:
y_train = train.pop("income") ## y_train은 pop쓰면 되겠군

In [22]:
y_train

0         >50K
1        <=50K
2        <=50K
3         >50K
4        <=50K
         ...  
29299    <=50K
29300    <=50K
29301    <=50K
29302    <=50K
29303    <=50K
Name: income, Length: 29304, dtype: object

원핫 인코딩 

In [23]:
train_oh = pd.get_dummies(train)
test_oh = pd.get_dummies(test)
print(train.shape, train_oh.shape, test.shape, test_oh.shape )

(29304, 15) (29304, 106) (3257, 15) (3257, 102)


-> 컬럼수 불일치 (train은 106개, test은 102개)

해결방법!! 
- 1.데이터 합치고 2.원핫인코딩 3. train, test분리
- 그러나 data leakage 문제가 있음
- 멘토님이 알려주신 방법 써야할듯. 

라벨인코더
- select_dtypes()

In [28]:
cols = train.select_dtypes(include='object').columns
cols

Index(['workclass', 'education', 'marital.status', 'occupation',
       'relationship', 'race', 'sex', 'native.country'],
      dtype='object')

In [30]:
from sklearn.preprocessing import LabelEncoder
cols = train.select_dtypes(include='object').columns
for col in cols:
    le = LabelEncoder()
    train[col] = le.fit_transform(train[col])
    test[col] = le.transform(test[col])
    
train.head()

,id,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country
0,3331,34.0,6,177331,15,10,2,9,0,2,1,4386,0,40.0,38
1,19749,58.0,3,290661,11,9,2,2,0,4,1,0,0,40.0,38
2,1157,48.0,3,125933,15,10,6,3,4,2,0,0,1669,38.0,38
3,693,58.0,3,100313,15,10,2,10,0,4,1,0,1902,40.0,38
4,12522,41.0,3,195661,15,10,2,13,0,4,1,0,0,54.0,38


## scaling

In [32]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29304 entries, 0 to 29303
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              29304 non-null  int64  
 1   age             29292 non-null  float64
 2   workclass       29304 non-null  int64  
 3   fnlwgt          29304 non-null  int64  
 4   education       29304 non-null  int64  
 5   education.num   29304 non-null  int64  
 6   marital.status  29304 non-null  int64  
 7   occupation      29304 non-null  int64  
 8   relationship    29304 non-null  int64  
 9   race            29304 non-null  int64  
 10  sex             29304 non-null  int64  
 11  capital.gain    29304 non-null  int64  
 12  capital.loss    29304 non-null  int64  
 13  hours.per.week  29291 non-null  float64
 14  native.country  29304 non-null  int64  
dtypes: float64(2), int64(13)
memory usage: 3.4 MB


In [41]:
# scaling 적용할 수치형 컬럼명 리스트로 만들기
num_cols = ['age','fnlwgt','education.num','capital.gain','capital.loss','hours.per.week']
#num_cols = num_cols.drop('id')
num_cols

['age',
 'fnlwgt',
 'education.num',
 'capital.gain',
 'capital.loss',
 'hours.per.week']

In [42]:
def get_data():
    train_copy = train.copy()
    test_copy = test.copy()
    return train_copy, test_copy

min_max (0~1)
- standard 은 평균 0 , 표편1 
- robust 은 각 값에서 중앙값 빼고 IQR로 나눔. (이상치 영향 덜받음)

In [43]:
train_copy, test_copy = get_data()

In [45]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
display(train_copy[num_cols].head())

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
0,34.0,177331,10,4386,0,40.0
1,58.0,290661,9,0,0,40.0
2,48.0,125933,10,0,1669,38.0
3,58.0,100313,10,0,1902,40.0
4,41.0,195661,10,0,0,54.0


In [46]:
train_copy[num_cols] = scaler.fit_transform(train_copy[num_cols])
test_copy[num_cols] = scaler.transform(test_copy[num_cols])

In [47]:
display(train_copy[num_cols].head())
display(test_copy[num_cols].head())

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
0,0.562500,0.112092,0.600000,0.04386,0.000000,0.397959
1,0.750000,0.189060,0.533333,0.00000,0.000000,0.397959
2,0.671875,0.077184,0.600000,0.00000,0.383150,0.377551
3,0.750000,0.059785,0.600000,0.00000,0.436639,0.397959
4,0.617188,0.124541,0.600000,0.00000,0.000000,0.540816


,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
0,0.601562,0.069118,0.800000,0.0,0.0,0.397959
1,0.593750,0.164239,0.600000,0.0,0.0,0.397959
2,0.640625,0.029278,0.533333,0.0,0.0,NaN
3,0.664062,0.010697,0.800000,0.0,0.0,0.500000
4,0.781250,0.118394,0.533333,0.0,0.0,0.397959


#### Q. target변수가 문자일때 인코딩? 
- 타겟값이 <=50k, >50k로 문자다 .
이때
- 이진분류 -> 인코딩 (map, replace 이용)
- 다중분류 -> 그대로 한다고함

## data split

In [48]:
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(train, y_train, test_size=0.2, random_state=42)


## 모델 학습, 평가

모델 학습

In [49]:
from sklearn.ensemble import RandomForestClassifier

rf= RandomForestClassifier(random_state=0)
rf.fit(X_train ,y_train)
pred = rf.predict_proba(X_val)


In [52]:
print(rf.classes_)
pred[:10] ## 확률이 반환

['<=50K' '>50K']


array([[0.17, 0.83],
       [1.  , 0.  ],
       [0.9 , 0.1 ],
       [0.98, 0.02],
       [1.  , 0.  ],
       [0.88, 0.12],
       [0.64, 0.36],
       [0.98, 0.02],
       [0.84, 0.16],
       [0.63, 0.37]])

- predict_proba 는 각 클래스 확률을 다 예측해줌( 이진분류니까 두개값이 생기는 셈) => ROC_AUC
- predict는 예측된 클래스만 반환 => 정확도, F1 score, 정밀도, 재현율

평가

In [57]:
c_pred = rf.predict(X_val)
from sklearn.metrics import f1_score
f1 = f1_score(y_val,c_pred, pos_label = ">50K")
f1

0.6775573706728899

In [59]:
pred

array([[0.17, 0.83],
       [1.  , 0.  ],
       [0.9 , 0.1 ],
       ...,
       [0.85, 0.15],
       [0.98, 0.02],
       [0.8 , 0.2 ]])

## submit
- 50,000 이상 확률값 묻는 중 >50 확률을 반환

In [61]:
rf.classes_ #두번째꺼 반환해야하는군

array(['<=50K', '>50K'], dtype=object)

In [62]:
submit = pd.DataFrame({'pred':pred[:,1]})
submit.to_csv("result.csv",index=False)

In [ ]:
submit

,pred
0,0.83
1,0.00
2,0.10
3,0.02
4,0.00
...,...
5856,0.00
5857,0.03
5858,0.15
5859,0.02
